### DISCUSSION TITLE
Errors on LAMMPS-CPU implementation

### DISCUSSION THREAD

**Discussion by bapatist** [ACEsuit/mace#161] [Q&A] [answered]
Hi all, 
trying to run my `model-lammps.pt` model on LAMMPS, I run into the following problems:
1. Using `ntasks=1` the `Segmentation Fault` error. (maybe this is something I don't want to use anyway)
[sub_ntasks1.txt](https://github.com/ACEsuit/mace/files/12648063/sub_ntasks1.txt)
[out.log](https://github.com/ACEsuit/mace/files/12648649/out.log)

2. With `ntasks=72` and `pair_style mace no_domain_decomposition`, I get a `RuntimeError (index is out of bound)` error from the torschscript interpreter.
[sub.txt](https://github.com/ACEsuit/mace/files/12648090/sub.txt) 
[err.txt](https://github.com/ACEsuit/mace/files/12648632/err.txt)
[out.log](https://github.com/ACEsuit/mace/files/12648104/out.log)

3. Using a plain `pair_style mace` with `ntasks>1` runs for a bit longer before crashing with out-of-memory kill events.
[sub2.txt](https://github.com/ACEsuit/mace/files/12648131/sub2.txt) 
[err.txt](https://github.com/ACEsuit/mace/files/12648591/err.txt)
[out.txt](https://github.com/ACEsuit/mace/files/12648139/out.txt) 

And the LAMMPS input file looks like this with `no_domain_decomposition` included for the second case
[lammps.txt](https://github.com/ACEsuit/mace/files/12648136/lammps.txt)

Please let me know if you see anything problematic, obvious or otherwise. Thanks!

**Comment by wcwitt** [comment, marked as answer]
Thanks for the detailed report! Let's stick with (1.) for now. Do you definitely need `atom_style full`?

**Comment by bapatist** [comment]
I don't necessarily need the full style. Retried with `atomic` style for the first case and I don't encounter a seg fault anymore and the simulation seems to be running. Trying out different options to improve speed now. I will write back here once I have more details. Thanks a lot!

**Comment by bapatist** [comment]
I was able to run NPT dynamics for 120 atoms (40 H2O) at 0.021 ns/day with `no_domain_decomposition` turned on. Attaching the log file here. Do you have suggestions to make this faster? [log.txt](https://github.com/ACEsuit/mace/files/12651206/log.txt)

I was trying to use `ntasks=72` with `no_domain_decomposition` turned off. Or is it not implemented to work like this yet? Attaching the error I get when I try to do this:


without `no_domain_decomposition`:  [err_no_nodomain.txt](https://github.com/ACEsuit/mace/files/12651280/err_no_nodomain.txt)
with `no_domain_decomposition`: [err_nodomain.txt](https://github.com/ACEsuit/mace/files/12651282/err_nodomain.txt)

**Comment by ilyes319** [comment]
Using GPUs instead of CPUs would be the most straight forward to make this faster. Even a quite cheap GPU would provide order of magnitude speed up (as MACE is a GPU code).

**Comment by wcwitt** [comment]
@ilyes319's answer will give you the best speedup, I'm fairly certain. But I'll try to address the CPU questions anyway.

> I was able to run NPT dynamics for 120 atoms (40 H2O) at 0.021 ns/day with no_domain_decomposition turned on. Attaching the log file here. Do you have suggestions to make this faster?

Make sure you are using ntask=1 with --exclusive, and then try different threading values using OMP_NUM_THREADS and MKL_NUM_THREADS. But it sounds like you were doing this already.

> I was trying to use ntasks=72 with no_domain_decomposition turned off

The problem here is that it loads 72 separate models and you run out of memory quickly. You could experiment with a few tasks per node, adjusting the threading accordingly, but the GPU will be better if you can use it.

Please do let us know how things go! If I think of anything else I will tell you.

**Comment by bapatist** [comment]
Hi! Just wanted to update on this: we have GPU models running with LAMMPS and indeed these are pretty fast! Thanks a lot.
